driver/controller.go

```diff
@@ -772,93 +772,119 @@ func (d *Driver) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequ
-		}
-	}
-
-	if nextToken != 0 && req.MaxEntries != 0 {
-		return nil, status.Errorf(codes.Aborted,
-			"ListSnapshots invalid arguments starting token: %d and max entries: %d can't be non null at the same time", nextToken, req.MaxEntries)
-	}
-
+	listResp := &csi.ListSnapshotsResponse{}
 	log := d.log.WithFields(logrus.Fields{
+		"snapshot_id":        req.SnapshotId,
+		"source_volume_id":   req.SourceVolumeId,
 		"req_starting_token": req.StartingToken,
 		"method":             "list_snapshots",
 	})
 	log.Info("list snapshots is called")
 
-	// fetch all entries
-	listOpts := &godo.ListOptions{
-		PerPage: int(req.MaxEntries),
-	}
-	var snapshots []godo.Snapshot
-	for {
-		snaps, resp, err := d.snapshots.ListVolume(ctx, listOpts)
+	if req.SnapshotId != "" {
+		snapshot, resp, err := d.snapshots.Get(ctx, req.SnapshotId)
 		if err != nil {
-			return nil, status.Errorf(codes.Aborted, "ListSnapshots listing volume snapshots has failed: %s", err.Error())
+			if resp == nil || resp.StatusCode != http.StatusNotFound {
+				return nil, status.Errorf(codes.Internal, "failed to get snapshot by ID %s: %s", req.SnapshotId, err)
+			}
+		} else {
+			snap, err := toCSISnapshot(snapshot)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal,
+					"failed to convert DO snapshot to CSI snapshot: %s", err)
+			}
+			listResp = &csi.ListSnapshotsResponse{
+				Entries: []*csi.ListSnapshotsResponse_Entry{
+					{
+						Snapshot: snap,
+					},
+				},
+			}
+		}
+	} else {
+		// Pagination in the CSI world works different than at DO. CSI sends the
+		// `req.MaxEntries` to indicate how much snapshots it wants. The
+		// req.StartingToken is returned by us, if we somehow need to indicate that
+		// we couldn't fetch and need to fetch again. But it's NOT the page number.
+		// I.e: suppose CSI wants us to fetch 50 entries, we only fetch 30, we need to
+		// return NextToken as 31 (so req.StartingToken will be set to 31 when CSI
+		// calls us again), to indicate that we want to continue returning from the
+		// index 31 up to 50.
+
+		var nextToken int
+		var err error
+		if req.StartingToken != "" {
+			nextToken, err = strconv.Atoi(req.StartingToken)
+			if err != nil {
+				return nil, status.Errorf(codes.Aborted, "ListSnapshots starting token %s is not valid : %s",
+					req.StartingToken, err.Error())
+			}
 		}
 
-		snapshots = append(snapshots, snaps...)
-
-		if resp.Links == nil || resp.Links.IsLastPage() {
-			break
+		if nextToken != 0 && req.MaxEntries != 0 {
+			return nil, status.Errorf(codes.Aborted,
+				"ListSnapshots invalid arguments starting token: %d and max entries: %d can't be non null at the same time", nextToken, req.MaxEntries)
 		}
 
-		page, err := resp.Links.CurrentPage()
-		if err != nil {
-			return nil, err
+		// fetch all entries
+		listOpts := &godo.ListOptions{
+			PerPage: int(req.MaxEntries),
 		}
+		var snapshots []godo.Snapshot
+		for {
+			snaps, resp, err := d.snapshots.ListVolume(ctx, listOpts)
```

Comment:
I'm a little confused with this pagination section below. It seems like we paginate through every entry from the start, ignoring the offset of `NextToken` that is returned. Can we just determine the "pageNumber" in DO pagination terms as a calculation of the pageCount and the offset?

Reply:
I think you're right, the logic looks off. I did not pay much attention to it because existing code was just moved into an `else` branch but was otherwise not modified. So I believe the pagination had been broken before. Great catch. 👍 

If you don't mind then I'd like to address this issue in a separate PR. FWIW, the flawed pagination should not have impacted users much since csi-snapshotter always passes in a snapshot ID in the `ListSnapshots` request, and our handling of that parameter is only fixed by this parameter. Being good CSI citizens, however, we should still fix the pagination for the case that `ListSnapshots` is ever consumed exhaustively either in the future or by other COs (container orchestrators) today already.

Reply:
:+1: That works for me.

Reply:
Fix submitted in #300.